File: code/scenes/src/main/java/org/fbme/scenes/ext/contextMenu/CellSelectionOnPopupTrigger.java
package org.fbme.scenes.ext.contextMenu;

import jetbrains.mps.nodeEditor.selection.EditorCellLabelSelection;
import jetbrains.mps.openapi.editor.EditorComponent;
import jetbrains.mps.openapi.editor.cells.CellConditions;
import jetbrains.mps.openapi.editor.cells.CellTraversalUtil;
import jetbrains.mps.openapi.editor.cells.EditorCell;
import jetbrains.mps.openapi.editor.extensions.EditorExtension;
import jetbrains.mps.openapi.editor.selection.Selection;
import jetbrains.mps.openapi.editor.selection.SelectionManager;
import jetbrains.mps.openapi.editor.selection.SingularSelection;
import org.jetbrains.annotations.NotNull;

import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.util.HashMap;
import java.util.Map;

public class CellSelectionOnPopupTrigger {

    private static final Map<EditorComponent, CellSelectionOnPopupTrigger> INSTANCES = new HashMap<>();

    public static final EditorExtension EDITOR_EXTENSION = new EditorExtension() {
        public boolean isApplicable(@NotNull EditorComponent ec) {
            return ec instanceof jetbrains.mps.nodeEditor.EditorComponent;
        }

        public void install(@NotNull EditorComponent ec) {
            CellSelectionOnPopupTrigger inst = new CellSelectionOnPopupTrigger((jetbrains.mps.nodeEditor.EditorComponent) ec);
            inst.install();
            INSTANCES.put(ec, inst);
        }

        public void uninstall(@NotNull EditorComponent ec) {
            INSTANCES.remove(ec).uninstall();
        }
    };


    private final jetbrains.mps.nodeEditor.EditorComponent myEditorComponent;
    private final MouseAdapter myMouseListener = new MouseAdapter() {
        @Override
        public void mousePressed(MouseEvent event) {
            if (!(event.isPopupTrigger())) {
                return;
            }

            SelectionManager selectionManager = myEditorComponent.getSelectionManager();
            if (isTrivialSelection(selectionManager.getSelection())) {

                EditorCell leaf = myEditorComponent.getRootCell().findLeaf(event.getX(), event.getY());
                if (leaf != null && CellTraversalUtil.getFoldedParent(leaf) != null) {
                    return;
                }
                if (leaf == null || !leaf.isSelectable()) {
                    leaf = myEditorComponent.getRootCell().findNearestLeafOnLine(event.getX(), event.getY(), CellConditions.SELECTABLE);
                }
                selectionManager.setSelection(leaf);
            }
        }
    };

    private boolean isTrivialSelection(@NotNull Selection selection) {
        if (selection instanceof EditorCellLabelSelection) {
            EditorCellLabelSelection labelSelection = (EditorCellLabelSelection) selection;
            return !(labelSelection.hasNonTrivialSelection());
        }
        return selection instanceof SingularSelection;
    }

    public CellSelectionOnPopupTrigger(@NotNull jetbrains.mps.nodeEditor.EditorComponent ec) {
        myEditorComponent = ec;

    }

    private void install() {
        MouseListener[] mouseListeners = myEditorComponent.getMouseListeners();
        for (MouseListener ml : mouseListeners) {
            myEditorComponent.removeMouseListener(ml);
        }
        myEditorComponent.addMouseListener(myMouseListener);
        for (MouseListener ml : mouseListeners) {
            myEditorComponent.addMouseListener(ml);
        }
    }

    private void uninstall() {
        myEditorComponent.removeMouseListener(myMouseListener);
    }

}

File: code/scenes/src/main/kotlin/org/fbme/scenes/ext/contextMenu/CellSelectionOnPopupTrigger.kt
package org.fbme.scenes.ext.contextMenu

import jetbrains.mps.nodeEditor.EditorComponent
import jetbrains.mps.nodeEditor.selection.EditorCellLabelSelection
import jetbrains.mps.openapi.editor.cells.CellConditions
import jetbrains.mps.openapi.editor.cells.CellTraversalUtil
import jetbrains.mps.openapi.editor.extensions.EditorExtension
import jetbrains.mps.openapi.editor.selection.Selection
import jetbrains.mps.openapi.editor.selection.SingularSelection
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.SwingUtilities

class CellSelectionOnPopupTrigger(private val editorComponent: EditorComponent) {
    private val mouseListener: MouseAdapter = object : MouseAdapter() {
        override fun mousePressed(event: MouseEvent) {
            if (!(event.isPopupTrigger || SwingUtilities.isRightMouseButton(event))) {
                return
            }
            val selectionManager = editorComponent.selectionManager
            val selection = selectionManager.selection
            if (selection != null && isTrivialSelection(selection!!)) {
                var leaf = editorComponent.rootCell.findLeaf(event.x, event.y)
                if (leaf != null && CellTraversalUtil.getFoldedParent(leaf) != null) {
                    return
                }
                if (leaf == null || !leaf.isSelectable) {
                    leaf = editorComponent.rootCell.findNearestLeafOnLine(event.x, event.y, CellConditions.SELECTABLE)
                }
                selectionManager.setSelection(leaf)
            }
        }
    }

    private fun isTrivialSelection(selection: Selection): Boolean {
        if (selection is EditorCellLabelSelection) {
            return !selection.hasNonTrivialSelection()
        }
        return selection is SingularSelection
    }

    private fun install() {
        val mouseListeners = editorComponent.mouseListeners
        for (ml in mouseListeners) {
            editorComponent.removeMouseListener(ml)
        }
        editorComponent.addMouseListener(mouseListener)
        for (ml in mouseListeners) {
            editorComponent.addMouseListener(ml)
        }
    }

    private fun uninstall() {
        editorComponent.removeMouseListener(mouseListener)
    }

    companion object {
        private val INSTANCES: MutableMap<jetbrains.mps.openapi.editor.EditorComponent, CellSelectionOnPopupTrigger> =
            HashMap()
        val EDITOR_EXTENSION: EditorExtension = object : EditorExtension {
            override fun isApplicable(ec: jetbrains.mps.openapi.editor.EditorComponent): Boolean {
                return ec is EditorComponent
            }

            override fun install(ec: jetbrains.mps.openapi.editor.EditorComponent) {
                val inst = CellSelectionOnPopupTrigger(ec as EditorComponent)
                inst.install()
                INSTANCES[ec] = inst
            }

            override fun uninstall(ec: jetbrains.mps.openapi.editor.EditorComponent) {
                INSTANCES.remove(ec)!!.uninstall()
            }
        }
    }
}
